modules/axtask/src/run_queue.rs

```diff
@@ -176,65 +431,98 @@ impl AxRunQueue {
             return;
         }
 
+        // Claim the task as running, we do this before switching to it
+        // such that any running task will have this set.
+        #[cfg(feature = "smp")]
+        next_task.set_on_cpu(true);
+
         unsafe {
             let prev_ctx_ptr = prev_task.ctx_mut_ptr();
             let next_ctx_ptr = next_task.ctx_mut_ptr();
 
+            // Store the weak pointer of **prev_task** in **next_task**'s struct.
+            #[cfg(feature = "smp")]
+            next_task.set_prev_task(prev_task.clone());
+
             // The strong reference count of `prev_task` will be decremented by 1,
             // but won't be dropped until `gc_entry()` is called.
             assert!(Arc::strong_count(prev_task.as_task_ref()) > 1);
             assert!(Arc::strong_count(&next_task) >= 1);
 
             CurrentTask::set_current(prev_task, next_task);
+
             (*prev_ctx_ptr).switch_to(&*next_ctx_ptr);
+
+            // Current it's **next_task** running on this CPU, clear the `prev_task`'s `on_cpu` field
+            // to indicate that it has finished its scheduling process and no longer running on this CPU.
+            #[cfg(feature = "smp")]
+            crate::current().clear_prev_task_on_cpu();
         }
     }
 }
 
 fn gc_entry() {
     loop {
         // Drop all exited tasks and recycle resources.
-        let n = EXITED_TASKS.lock().len();
+        let n = EXITED_TASKS.with_current(|exited_tasks| exited_tasks.len());
         for _ in 0..n {
             // Do not do the slow drops in the critical section.
-            let task = EXITED_TASKS.lock().pop_front();
+            let task = EXITED_TASKS.with_current(|exited_tasks| exited_tasks.pop_front());
             if let Some(task) = task {
                 if Arc::strong_count(&task) == 1 {
                     // If I'm the last holder of the task, drop it immediately.
                     drop(task);
                 } else {
                     // Otherwise (e.g, `switch_to` is not compeleted, held by the
                     // joiner, etc), push it back and wait for them to drop first.
-                    EXITED_TASKS.lock().push_back(task);
+                    EXITED_TASKS.with_current(|exited_tasks| exited_tasks.push_back(task));
                 }
             }
         }
-        WAIT_FOR_EXIT.wait();
+        unsafe { WAIT_FOR_EXIT.current_ref_raw() }.wait();
```

Comment:
Explain why use `current_ref_raw`

Reply:
I think `with_current` with preemption disabled is safer here

Reply:
Nononono, `current_ref_raw` is required here, we cannot block current task with preemption disabled, so we use `current_ref_raw` to get the `WAIT_FOR_EXIT`'s reference here to avoid the use of `NoPreemptGuard`.

Reply:
use `with_current` will lead to https://github.com/arceos-org/arceos/actions/runs/11287770863/job/31394266857?pr=176

Reply:
Do we need to add `kernel_guard::IrqSave` here?
@Azure-stars

Reply:
Since gc task is pinned to the current CPU, there is no affection if the gc task is preempted during the process.
So `kernel_guard::IrqSave` is not needed.